### ISSUE TITLE
CosyVoice3 load_trt=True + fp16=True produces non-finite (NaN) audio in 64/64 generations on the official Fun-CosyVoice3-0.5B-2512 checkpoint

### ISSUE THREAD

**Issue by shuvro-sarker** [QwenAudio/CosyVoice#1930]
**Type:** Bug report (correctness, not performance)

**Relates to:** #1822 (closed by stale-bot, unresolved/disputed) and #1699 (closed by stale-bot, unresolved) both report the same or an adjacent symptom on the same warning message. This report is not the first to notice it, but it quantifies the failure with a controlled, seed-matched, repeatable corpus alongside fp32-TensorRT and eager controls — and the numbers say the upstream warning undersells a 100% correctness failure as a performance caveat.

## Environment

- Model: `Fun-CosyVoice3-0.5B-2512` (official, unmodified)
- Config: `AutoModel(model_dir=..., load_trt=True, fp16=True)`
- GPU: NVIDIA L40S (46 GB)
- Repo state checked: `main` @ `074ca6dc9e80a2f424f1f74b48bdd7d3fea531cc`

## Code path

`cosyvoice/cli/cosyvoice.py`, `class CosyVoice3(CosyVoice2).__init__`:

```python
if load_trt:
    if self.fp16 is True:
        logging.warning('DiT tensorRT fp16 engine have some performance issue, use at caution!')
    self.model.load_trt(
        '{}/flow.decoder.estimator.{}.mygpu.plan'.format(model_dir, 'fp16' if self.fp16 is True else 'fp32'),
        ..., self.fp16)
```

The engine is compiled from `flow.decoder.estimator.fp16.onnx` into a cached `.mygpu.plan`.

## Minimal reproduction

```python
from cosyvoice.cli.cosyvoice import CosyVoice3
from cosyvoice.utils.file_utils import load_wav
import torch

model = CosyVoice3('Fun-CosyVoice3-0.5B-2512', load_trt=True, fp16=True)
prompt_wav = load_wav('english_prompt.wav', 16000)

for text in corpus:  # 16 utterances
    for trial in range(4):
        for chunk in model.inference_zero_shot(text, prompt_text, prompt_wav, stream=True):
            audio = chunk['tts_speech']
            assert torch.isfinite(audio).all(), "non-finite sample in output"
```

Run against a 16-utterance English conversational-turn corpus × 4 trials = 64 generations, with identical seeds to the fp32-TensorRT and eager control arms in the same sweep.

## Observed vs expected

- **Expected:** either finite audio, or a hard failure that stops the pipeline from silently shipping bad audio — not a log-level warning describing it as a "performance issue."
- **Observed:** **64/64 generations produced non-finite (NaN) samples** — a 0% pass rate on a `torch.isfinite` check, versus 0 failures on the fp32-TensorRT and eager-fp32 controls run in the same sweep with identical inputs and seeds.
- **The warning text is misleading about severity.** "have some performance issue, use at caution" reads as a latency/quality tradeoff. In this configuration it is a total correctness failure — the output is not degraded audio, it is not audio at all. Someone tuning purely on latency would not notice: in the same sweep, the fp16-TensorRT rung measured **12% faster than the eager fp32 baseline at p99** (2628 ms vs 2738 ms), which is exactly the kind of result that passes a latency-only benchmark and ships.

## Measured impact

Same 16-utterance × 4-trial corpus, same seeds, L40S:

| rung | config | finite-audio check | p99 first-packet |
|---|---|---|---|
| trt-fp16 | `load_trt=True, fp16=True` | **64/64 fail** (non-finite) | 2628 ms (12% faster than eager — meaningless given the failure) |
| trt-fp32 | `load_trt=True, fp16=False` | 0/64 fail | 2434 ms |
| eager | no TensorRT | 0/64 fail | 2738 ms |

Because every generation on the fp16 rung failed the finite-sample check, no ASR/WER scoring was performed on it.

## Related upstream reports (not resolved, not root-caused)

- **#1822** — identical repro (`CosyVoice3(..., load_trt=True, load_vllm=True, fp16=True)` on `Fun-CosyVoice3-0.5B`) and identical symptom (NaN on save / silent WAV). One commenter on the same GPU family (L40) claims pinning `tensorrt-cu12==10.13.3.9` resolves it; another directly contradicts that in the same thread ("fp16全是NAN"). Auto-closed by the stale-bot with no maintainer follow-up. **We have not tested the `tensorrt-cu12==10.13.3.9` pin ourselves** — flagging it as an unverified lead from the same GPU family, not a confirmed fix.
- **#1699** — asked upstream to explain the warning message directly; two speculative replies, no data, no fix. Auto-closed by the stale-bot.
- **#1173** (open) — same NaN/silence symptom under TensorRT FP16, but on a self-trained flow2 checkpoint; related mechanism, different scope, so kept separate rather than merged into this report.

**This report's contribution beyond the above:** a controlled, seed-matched, 64-generation repro specifically on the official, unmodified `Fun-CosyVoice3-0.5B-2512` checkpoint, run alongside fp32-TensorRT and eager controls that both pass cleanly on the same inputs — which narrows the defect to the FP16 TensorRT engine specifically, not to TensorRT in general, the checkpoint's training, or the reference audio.

## Suggested resolution (any of)

1. Fix the FP16 engine path (root cause in the ODE-solve numerics or the ONNX export).
2. Until then, fail hard instead of warning — `raise` rather than `logging.warning` — so the configuration cannot silently ship non-finite audio.
3. At minimum, change the warning text to state it is a correctness failure, and document `load_trt=True, fp16=False` as the supported TensorRT configuration (that is what our deployment uses; it is both correct and faster than eager).

**Comment by aluminumbox** [comment]
我是CosyVoice自动答疑助手（机器人自动回复，非人工）。

感谢您提供了如此详尽、可复现的 bug report，包括对照组数据和种子匹配的设计，这对定位问题非常有价值。

## 关于此问题的分析

您的发现与之前 #1822 和 #1699 报告的现象一致：在使用 `load_trt=True, fp16=True` 配置时，FP16 TensorRT 引擎在 ODE solver（flow matching decoder）中产生数值溢出（NaN）。您的对照实验进一步确认：

1. **问题定位在 FP16 TensorRT 引擎本身**，而非 TensorRT 框架、checkpoint 或参考音频
2. FP32 TensorRT（`load_trt=True, fp16=False`）在相同条件下完全正常且性能更优（2434ms vs eager 2738ms）

## 临时规避方案

在官方修复前，推荐使用以下配置：

```python
model = CosyVoice3('Fun-CosyVoice3-0.5B-2512', load_trt=True, fp16=False)
```

即 **TensorRT FP32 模式**——根据您的测试数据，它既保证了输出正确性（0/64 失败），又比 eager 模式快约 11%。

## 关于 #1822 中提到的 tensorrt 版本 pin

有用户反馈 `tensorrt-cu12==10.13.3.9` 可能有帮助，但也有用户在相同 GPU 系列上报告该 pin 无效。如您有条件，可尝试验证，但建议以 fp32 TensorRT 作为稳定方案。

## 关于您提出的修复建议

您的三个建议（修复根因 / 改为 raise 而非 warning / 至少修正 warning 文案）都很合理。已将此问题记录并标记给维护者关注。特别是将 warning 文案从 "performance issue" 修正为明确的 correctness failure 提示，应是一个低成本高收益的改进。

---
以上信息基于现有 issue 讨论和您提供的测试数据，以仓库最新代码为准。维护者会进一步跟进此问题。
